### PR TITLE
Make OptimizedModule state_dict use public keys

### LINKED ISSUE #149894: [ued][f5-tts][dynamo] `torch.compile` changes state dict
### 🐛 Describe the bug

Applying torch.compile to a model changes the state dict, and breaking loading of existing state dict checkpoints.

This issue is to figure out how to instruct users on avoiding this issue.

### Error logs

_No response_

### Versions

N/A

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.16.0) (oldest at bottom):
* __->__ #185752

`torch.compile(module)` returns an `OptimizedModule` wrapper that registers the original module as `_orig_mod`. Because the wrapper previously inherited the default `nn.Module.state_dict` and `load_state_dict` traversal, checkpoint keys exposed that implementation detail as `_orig_mod.*`. That changed the keyspace relative to eager modules and broke loading checkpoints across compiled and uncompiled versions.

Fix this by making `OptimizedModule.state_dict()` save the wrapped module's public keyspace while preserving wrapper state_dict hooks. For load, keep the normal `nn.Module.load_state_dict` traversal and translate public keys to the wrapper's internal `_orig_mod.` path inside `OptimizedModule._load_from_state_dict`. This keeps compiled modules compatible with eager checkpoints and avoids bypassing normal load hook ordering.

The load path also accepts old compiled checkpoints that already contain `_orig_mod.` keys. The compatibility path is guarded by keyspace scoring so a real user child named `_orig_mod` keeps its public keys, and ambiguous partial keyspaces stay public rather than being silently reinterpreted. Metadata is copied before translation so caller state_dict metadata is not mutated, and the private load postprocess hook maps internal incompatible key names back to public names before user-visible reporting.

I considered a top-level `OptimizedModule.load_state_dict` delegation to `_orig_mod`, but that skipped wrapper load hooks and raised before post-hooks on load errors. Keeping the base traversal and doing prefix translation at `_load_from_state_dict` matches `nn.Module` semantics more closely, including nested compiled modules.

Fixes #149894
Generated by my agent

Test Plan:
- `python test/dynamo/test_modules.py OptimizedModuleTest -v`
- `python test/nn/test_load_state_dict.py -v`
- `python test/nn/test_module_hooks.py -k load_state_dict -v`
- `python test/dynamo/test_compile.py InPlaceCompilationTests.test_state_dict_save -v`
- `lintrunner -a`

Benchmark Results:
- Benchmark: sequential model with 512 `torch.nn.Linear(8, 8)` children, one state_dict with metadata, 5 warmup loads, 40 measured `model.load_state_dict(state)` repeats.
- Baseline `HEAD`: median 47.436319 ms, mean 47.581416 ms, min 43.160838 ms, max 56.507357 ms.
- Current patch: median 44.562735 ms, mean 44.853332 ms, min 43.546528 ms, max 48.612572 ms.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I updated the PR to address the actionable CI failure and review feedback. The JIT failure in TestModuleAPIs.test_customized_state_dict_methods was caused by ScriptModule rejecting the new private load-state-dict postprocess helper, so I allowlisted that internal helper for RecursiveScriptModule. I also added the fast path suggested in review so ordinary public-key loads skip the legacy _orig_mod keyspace traversal. The docs failures were aws: command not found during doc preview upload, which appears to be CI environment/tooling rather than this patch. Local verification: targeted JIT state_dict tests, targeted OptimizedModule tests, full OptimizedModuleTest, and lintrunner -a.

**Comment by jansel** [comment]
My agent says I rebased this PR onto current main and addressed the remaining compatibility gaps found during review. Descendant load hooks now see public incompatible-key names across siblings and nested compiled modules; ambiguous legacy keyspaces support checkpoints produced only by state-dict hooks; and wrappers with their own persistent state, including state registered by pre-hooks, retain legacy traversal so values and colliding names are not dropped. The suggested torch.compile(torch.compile(module)) case was also checked: the second compile intentionally unwraps to a callable without a state_dict surface, while nested compiled-module state is covered directly. Local verification passed 84 OptimizedModule tests, 29 shared load-state tests, 8 load-hook tests, 2 JIT state-dict tests, the in-place compile checkpoint test, and lintrunner -a.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says this PR is being closed as part of a cleanup of open agentic PRs to focus review bandwidth on a smaller set; the linked issue remains open.
